session.go

```diff
@@ -1511,46 +1597,34 @@ func (s *session) RefreshTxnCtx(goCtx goctx.Context) error {
 
 // ActivePendingTxn implements Context.ActivePendingTxn interface.
 func (s *session) ActivePendingTxn() error {
-	if s.Transaction != nil && s.Transaction.Valid() {
+	if s.txn.Valid() {
 		return nil
 	}
-	if s.txnFuture == nil {
-		return errors.New("transaction future is not set")
-	}
-	future := s.txnFuture
-	s.txnFuture = nil
-
-	txn, err := future.wait()
-	if err != nil {
+	// The transaction status should be TxnStateFuture.
+	if err := s.txn.changePendingToValid(); err != nil {
 		return errors.Trace(err)
 	}
-	txn.SetCap(s.getMembufCap())
```

Comment:
why remove this line?

Reply:
Unnecessary, the underlying membuf in `StateTxn` is reused, so we don't need to pre-alloc the membuf.
B.T.W, I'll remove the BufStore in WriteStmtBufs because membuf in StateTxn can replace it.

Reply:
we prealloc it just for importing data.

Reply:
Now that the StateTxn already contains a buffer, we can make full use of it.
Importing data also benefit from this.

Reply:
NO, txn.SetCap is setting capability to TxnState.Transaction's membuf, not TxnState.buf's cap, because the we new a TxnState when every transaction begin, so we still need to set cap to reduce allocations.

Reply:
Done.